File: android-java/app/src/main/java/io/agora/convoai/example/startup/ui/AgentChatActivity.java
package io.agora.convoai.example.startup.ui;

import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import io.agora.convoai.convoaiApi.AgentState;
import io.agora.convoai.convoaiApi.Transcript;
import io.agora.convoai.convoaiApi.TranscriptStatus;
import io.agora.convoai.convoaiApi.TranscriptType;
import io.agora.convoai.example.startup.R;
import io.agora.convoai.example.startup.databinding.ActivityAgentChatBinding;
import io.agora.convoai.example.startup.databinding.ItemTranscriptAgentBinding;
import io.agora.convoai.example.startup.databinding.ItemTranscriptUserBinding;
import io.agora.convoai.example.startup.ui.common.BaseActivity;
import io.agora.convoai.example.startup.ui.CommonDialog;
import io.agora.convoai.example.startup.tools.PermissionHelp;

import java.util.ArrayList;
import java.util.List;

/**
 * Activity for agent chat interface
 * Layout: log, agent status, transcript, start/control buttons
 */
public class AgentChatActivity extends BaseActivity<ActivityAgentChatBinding> {

    private AgentChatViewModel viewModel;
    private PermissionHelp mPermissionHelp;
    private TranscriptAdapter transcriptAdapter;

    // Track whether to automatically scroll to bottom
    private boolean autoScrollToBottom = true;
    private boolean isScrollBottom = false;

    @Override
    protected ActivityAgentChatBinding getViewBinding() {
        return ActivityAgentChatBinding.inflate(getLayoutInflater());
    }

    @Override
    public void initData() {
        super.initData();
        viewModel = new ViewModelProvider(this).get(AgentChatViewModel.class);
        mPermissionHelp = new PermissionHelp(this);
        transcriptAdapter = new TranscriptAdapter();

        // Observe UI state changes
        observeUiState();

        // Observe transcript list changes
        observeTranscriptList();

        // Observe debug log changes
        observeDebugLogs();
    }

    @Override
    protected void initView() {
        ActivityAgentChatBinding binding = getBinding();
        if (binding == null) {
            return;
        }

        setOnApplyWindowInsetsListener(binding.getRoot());

        // Setup RecyclerView for transcript list
        setupRecyclerView();

        // Start button click listener
        binding.btnStart.setOnClickListener(v -> {
            // Generate random channel name each time joining channel
            String channelName = AgentChatViewModel.generateRandomChannelName();

            // Check microphone permission before joining channel
            checkMicrophonePermission(granted -> {
                if (granted) {
                    viewModel.joinChannelAndLogin(channelName);
                } else {
                    Toast.makeText(
                            AgentChatActivity.this,
                            "Microphone permission is required to join channel",
                            Toast.LENGTH_LONG
                    ).show();
                }
            });
        });

        // Mute button click listener
        binding.btnMute.setOnClickListener(v -> viewModel.toggleMute());

        // Stop button click listener
        binding.btnStop.setOnClickListener(v -> viewModel.hangup());
    }

    @Override
    public boolean supportOnBackPressed() {
        return false;
    }

    private void checkMicrophonePermission(PermissionCallback callback) {
        if (mPermissionHelp.hasMicPerm()) {
            callback.onResult(true);
        } else {
            mPermissionHelp.checkMicPerm(
                    () -> callback.onResult(true),
                    () -> {
                        showPermissionDialog(
                                "Permission Required",
                                "Microphone permission is required for voice chat. Please grant the permission to continue.",
                                result -> {
                                    if (result) {
                                        mPermissionHelp.launchAppSettingForMic(
                                                () -> callback.onResult(true),
                                                () -> callback.onResult(false)
                                        );
                                    } else {
                                        callback.onResult(false);
                                    }
                                }
                        );
                    }
            );
        }
    }

    private void showPermissionDialog(String title, String content, PermissionCallback onResult) {
        if (isFinishing() || isDestroyed() || getSupportFragmentManager().isStateSaved()) {
            return;
        }

        new CommonDialog.Builder()
                .setTitle(title)
                .setContent(content)
                .setPositiveButton("Retry", () -> onResult.onResult(true))
                .setNegativeButton("Exit", () -> onResult.onResult(false))
                .setCancelable(false)
                .build()
                .show(getSupportFragmentManager(), "permission_dialog");
    }

    /**
     * Callback interface for permission check results
     */
    private interface PermissionCallback {
        void onResult(boolean granted);
    }

    /**
     * Setup RecyclerView for transcript list
     */
    private void setupRecyclerView() {
        ActivityAgentChatBinding binding = getBinding();
        if (binding == null) {
            return;
        }

        RecyclerView recyclerView = binding.rvTranscript;
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setReverseLayout(false);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setAdapter(transcriptAdapter);
        recyclerView.setItemAnimator(null);
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                switch (newState) {
                    case RecyclerView.SCROLL_STATE_IDLE:
                        // Check if at bottom when scrolling stops
                        isScrollBottom = !recyclerView.canScrollVertically(1);
                        if (isScrollBottom) {
                            autoScrollToBottom = true;
                            isScrollBottom = true;
                        }
                        break;

                    case RecyclerView.SCROLL_STATE_DRAGGING:
                        // When user actively drags
                        autoScrollToBottom = false;
                        break;
                }
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                // Show button when scrolling up a significant distance
                if (dy < -50) {
                    if (recyclerView.canScrollVertically(1)) {
                        autoScrollToBottom = false;
                    }
                }
            }
        });
    }

    private void observeUiState() {
        viewModel.uiState.observe(this, state -> {
            ActivityAgentChatBinding binding = getBinding();
            if (binding == null || state == null) {
                return;
            }

            // Update button visibility based on connection state
            boolean isConnected = state.connectionState == AgentChatViewModel.ConnectionState.Connected;
            boolean isConnecting = state.connectionState == AgentChatViewModel.ConnectionState.Connecting;

            // Show/hide buttons
            binding.llStart.setVisibility(isConnected ? View.GONE : View.VISIBLE);
            binding.llControls.setVisibility(isConnected ? View.VISIBLE : View.GONE);

            // Update button loading state
            if (isConnecting) {
                binding.btnStart.setText("Starting...");
                binding.btnStart.setEnabled(false);
            } else {
                binding.btnStart.setText("Start Agent");
                binding.btnStart.setEnabled(true);
            }

            // Update mute button UI
            binding.btnMute.setImageResource(
                    state.isMuted ? R.drawable.ic_mic_off : R.drawable.ic_mic
            );
        });

        // Observe agent state
        viewModel.agentState.observe(this, agentState -> {
            ActivityAgentChatBinding binding = getBinding();
            if (binding == null) {
                return;
            }

            if (agentState != null) {
                // Update agent status text using state.value
                binding.tvAgentStatus.setText(agentState.getValue());
            } else {
                // Agent state is null, show default text
                binding.tvAgentStatus.setText("Unknown");
            }
        });
    }

    private void observeTranscriptList() {
        viewModel.transcriptList.observe(this, transcriptList -> {
            // Update transcript list
            transcriptAdapter.submitList(transcriptList != null ? transcriptList : new ArrayList<>());
            if (autoScrollToBottom) {
                scrollToBottom();
            }
        });
    }

    private void observeDebugLogs() {
        viewModel.debugLogList.observe(this, logList -> {
            ActivityAgentChatBinding binding = getBinding();
            if (binding == null) {
                return;
            }

            // Update log text with all logs, separated by newlines
            if (logList != null && !logList.isEmpty()) {
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < logList.size(); i++) {
                    if (i > 0) {
                        sb.append("\n");
                    }
                    sb.append(logList.get(i));
                }
                binding.tvLog.setText(sb.toString());
            } else {
                binding.tvLog.setText("log");
            }

            // Auto scroll to bottom
            binding.tvLog.post(() -> {
                View parent = (View) binding.tvLog.getParent();
                if (parent instanceof android.widget.ScrollView) {
                    ((android.widget.ScrollView) parent).fullScroll(View.FOCUS_DOWN);
                }
            });
        });
    }

    /**
     * Scroll RecyclerView to the bottom to show latest transcript
     */
    private void scrollToBottom() {
        ActivityAgentChatBinding binding = getBinding();
        if (binding == null) {
            return;
        }

        RecyclerView recyclerView = binding.rvTranscript;
        int lastPosition = transcriptAdapter.getItemCount() - 1;
        if (lastPosition < 0) {
            return;
        }

        recyclerView.stopScroll();
        LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
        if (layoutManager == null) {
            return;
        }

        // Use single post call to handle all scrolling logic
        recyclerView.post(() -> {
            layoutManager.scrollToPosition(lastPosition);

            // Handle extra-long messages that exceed viewport height
            View lastView = layoutManager.findViewByPosition(lastPosition);
            if (lastView != null && lastView.getHeight() > recyclerView.getHeight()) {
                int offset = recyclerView.getHeight() - lastView.getHeight();
                layoutManager.scrollToPositionWithOffset(lastPosition, offset);
            }

            isScrollBottom = true;
        });
    }
}

/**
 * Adapter for displaying transcript list with different view types for USER and AGENT
 */
class TranscriptAdapter extends ListAdapter<Transcript, RecyclerView.ViewHolder> {

    private static final int VIEW_TYPE_USER = 0;
    private static final int VIEW_TYPE_AGENT = 1;

    public TranscriptAdapter() {
        super(new TranscriptDiffCallback());
    }

    @Override
    public int getItemViewType(int position) {
        Transcript transcript = getItem(position);
        if (transcript == null) {
            return VIEW_TYPE_USER;
        }
        return transcript.getType() == TranscriptType.USER ? VIEW_TYPE_USER : VIEW_TYPE_AGENT;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        switch (viewType) {
            case VIEW_TYPE_USER:
                return new UserViewHolder(ItemTranscriptUserBinding.inflate(inflater, parent, false));

            case VIEW_TYPE_AGENT:
                return new AgentViewHolder(ItemTranscriptAgentBinding.inflate(inflater, parent, false));

            default:
                throw new IllegalArgumentException("Unknown view type: " + viewType);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        Transcript transcript = getItem(position);
        if (transcript == null) {
            return;
        }

        if (holder instanceof UserViewHolder) {
            ((UserViewHolder) holder).bind(transcript);
        } else if (holder instanceof AgentViewHolder) {
            ((AgentViewHolder) holder).bind(transcript);
        }
    }

    /**
     * ViewHolder for USER transcript items
     */
    static class UserViewHolder extends RecyclerView.ViewHolder {
        private final TextView tvType;
        private final TextView tvText;
        private final TextView tvStatus;

        UserViewHolder(ItemTranscriptUserBinding binding) {
            super(binding.getRoot());
            tvType = binding.tvTranscriptType;
            tvText = binding.tvTranscriptText;
            tvStatus = binding.tvTranscriptStatus;
        }

        void bind(Transcript transcript) {
            // Set transcript type badge with green color for USER
            tvType.setText("USER");
            android.graphics.drawable.Drawable drawable = ContextCompat.getDrawable(
                    itemView.getContext(), R.drawable.bg_type_badge);
            if (drawable != null) {
                drawable.setTint(Color.parseColor("#10B981"));
                tvType.setBackground(drawable);
            }

            // Set transcript text
            String text = transcript.getText();
            tvText.setText(text != null && !text.isEmpty() ? text : "(empty)");

            // Set transcript status with appropriate color
            String statusText;
            int statusColor;
            TranscriptStatus status = transcript.getStatus();
            if (status == TranscriptStatus.IN_PROGRESS) {
                statusText = "IN PROGRESS";
                statusColor = Color.parseColor("#FF9800");
            } else if (status == TranscriptStatus.END) {
                statusText = "END";
                statusColor = Color.parseColor("#4CAF50");
            } else if (status == TranscriptStatus.INTERRUPTED) {
                statusText = "INTERRUPTED";
                statusColor = Color.parseColor("#F44336");
            } else {
                statusText = "UNKNOWN";
                statusColor = Color.parseColor("#9E9E9E");
            }
            tvStatus.setText(statusText);
            tvStatus.setTextColor(statusColor);
        }
    }

    /**
     * ViewHolder for AGENT transcript items
     */
    static class AgentViewHolder extends RecyclerView.ViewHolder {
        private final TextView tvType;
        private final TextView tvText;
        private final TextView tvStatus;

        AgentViewHolder(ItemTranscriptAgentBinding binding) {
            super(binding.getRoot());
            tvType = binding.tvTranscriptType;
            tvText = binding.tvTranscriptText;
            tvStatus = binding.tvTranscriptStatus;
        }

        void bind(Transcript transcript) {
            // Set transcript type badge with indigo color for AGENT
            tvType.setText("AGENT");
            android.graphics.drawable.Drawable drawable = ContextCompat.getDrawable(
                    itemView.getContext(), R.drawable.bg_type_badge);
            if (drawable != null) {
                drawable.setTint(Color.parseColor("#6366F1"));
                tvType.setBackground(drawable);
            }

            // Set transcript text
            String text = transcript.getText();
            tvText.setText(text != null && !text.isEmpty() ? text : "(empty)");

            // Set transcript status with appropriate color
            String statusText;
            int statusColor;
            TranscriptStatus status = transcript.getStatus();
            if (status == TranscriptStatus.IN_PROGRESS) {
                statusText = "IN PROGRESS";
                statusColor = Color.parseColor("#FF9800");
            } else if (status == TranscriptStatus.END) {
                statusText = "END";
                statusColor = Color.parseColor("#4CAF50");
            } else if (status == TranscriptStatus.INTERRUPTED) {
                statusText = "INTERRUPTED";
                statusColor = Color.parseColor("#F44336");
            } else {
                statusText = "UNKNOWN";
                statusColor = Color.parseColor("#9E9E9E");
            }
            tvStatus.setText(statusText);
            tvStatus.setTextColor(statusColor);
        }
    }

    private static class TranscriptDiffCallback extends DiffUtil.ItemCallback<Transcript> {
        @Override
        public boolean areItemsTheSame(@NonNull Transcript oldItem, @NonNull Transcript newItem) {
            return oldItem.getTurnId() == newItem.getTurnId() && oldItem.getType() == newItem.getType();
        }

        @Override
        public boolean areContentsTheSame(@NonNull Transcript oldItem, @NonNull Transcript newItem) {
            return oldItem.equals(newItem);
        }
    }
}

File: feature-avatar/android-kotlin/app/src/main/java/io/agora/convoai/example/startup/ui/AgentChatActivity.kt
package io.agora.convoai.example.startup.ui

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import io.agora.convoai.example.startup.R
import io.agora.convoai.example.startup.databinding.ActivityAgentChatBinding
import io.agora.convoai.example.startup.databinding.ItemTranscriptAgentBinding
import io.agora.convoai.example.startup.databinding.ItemTranscriptUserBinding
import io.agora.convoai.example.startup.ui.common.BaseActivity
import io.agora.convoai.convoaiApi.Transcript
import io.agora.convoai.convoaiApi.TranscriptStatus
import io.agora.convoai.convoaiApi.TranscriptType
import kotlinx.coroutines.launch
import kotlin.text.ifEmpty
import androidx.core.graphics.toColorInt
import io.agora.convoai.example.startup.tools.PermissionHelp

/**
 * Activity for agent chat interface
 * Layout: log, agent status, transcript, start/control buttons
 */
class AgentChatActivity : BaseActivity<ActivityAgentChatBinding>() {

    private lateinit var viewModel: AgentChatViewModel
    private lateinit var mPermissionHelp: PermissionHelp
    private val transcriptAdapter: TranscriptAdapter = TranscriptAdapter()

    // Track whether to automatically scroll to bottom
    private var autoScrollToBottom = true
    private var isScrollBottom = false

    // Avatar video display state
    private var isAvatarExpanded = false
    private val avatarViewSmallWidth = 120  // dp
    private val avatarViewSmallHeight = 160  // dp

    override fun getViewBinding(): ActivityAgentChatBinding {
        return ActivityAgentChatBinding.inflate(layoutInflater)
    }

    override fun initData() {
        super.initData()
        viewModel = ViewModelProvider(this)[AgentChatViewModel::class.java]
        mPermissionHelp = PermissionHelp(this)
        
        // Load saved UIDs and fill input fields
        loadSavedUIDs()

        // Observe UI state changes
        observeUiState()

        // Observe transcript list changes
        observeTranscriptList()

        // Observe debug log changes
        observeDebugLogs()

        // Observe avatar joined state
        observeAvatarJoined()
    }

    override fun initView() {
        mBinding?.apply {
            setOnApplyWindowInsetsListener(root)

            // Setup RecyclerView for transcript list
            setupRecyclerView()

            // Setup ChannelInputView callback
            channelInputView.onJoinChannelListener = object : OnJoinChannelListener {
                override fun onJoinChannel(data: ChannelInputData) {
                    // Generate random channel name each time joining channel
                    val channelName = data.channelName.ifEmpty { 
                        AgentChatViewModel.generateRandomChannelName() 
                    }
                    
                    // Validate UIDs
                    if (data.userId == null || data.userId <= 0) {
                        viewModel.addStatusLog("ERROR: 用户UID不能为空")
                        return
                    }
                    if (data.agentUid == null || data.agentUid <= 0) {
                        viewModel.addStatusLog("ERROR: Agent UID不能为空")
                        return
                    }
                    if (data.avatarUid == null || data.avatarUid <= 0) {
                        viewModel.addStatusLog("ERROR: Avatar UID不能为空")
                        return
                    }

                    // Check microphone permission before joining channel
                    checkMicrophonePermission { granted ->
                        if (granted) {
                            viewModel.joinChannelAndLogin(channelName, data.userId, data.agentUid, data.avatarUid)
                        } else {
                            Toast.makeText(
                                this@AgentChatActivity,
                                "Microphone permission is required to join channel",
                                Toast.LENGTH_LONG
                            ).show()
                        }
                    }
                }
            }

            // Mute button click listener
            btnMute.setOnClickListener {
                viewModel.toggleMute()
            }

            // Stop button click listener
            btnStop.setOnClickListener {
                viewModel.hangup()
            }

            // Avatar video container click listener - expand/collapse
            avatarVideoContainer.setOnClickListener {
                val state = viewModel.uiState.value
                val isConnected = state.connectionState == AgentChatViewModel.ConnectionState.Connected
                if (isConnected) {
                    toggleAvatarSize()
                }
            }
        }
    }

    private fun checkMicrophonePermission(granted: (Boolean) -> Unit) {
        if (mPermissionHelp.hasMicPerm()) {
            granted.invoke(true)
        } else {
            mPermissionHelp.checkMicPerm(
                granted = { granted.invoke(true) },
                unGranted = {
                    showPermissionDialog(
                        "Permission Required",
                        "Microphone permission is required for voice chat. Please grant the permission to continue.",
                        onResult = {
                            if (it) {
                                mPermissionHelp.launchAppSettingForMic(
                                    granted = { granted.invoke(true) },
                                    unGranted = { granted.invoke(false) }
                                )
                            } else {
                                granted.invoke(false)
                            }
                        }
                    )
                }
            )
        }
    }

    private fun showPermissionDialog(title: String, content: String, onResult: (Boolean) -> Unit) {
        if (isFinishing || isDestroyed || supportFragmentManager.isStateSaved) return

        CommonDialog.Builder()
            .setTitle(title)
            .setContent(content)
            .setPositiveButton("Retry") {
                onResult.invoke(true)
            }
            .setNegativeButton("Exit") {
                onResult.invoke(false)
            }
            .setCancelable(false)
            .build()
            .show(supportFragmentManager, "permission_dialog")
    }

    private fun handleStop() {
        viewModel.hangup()
    }

    /**
     * Setup RecyclerView for transcript list
     */
    private fun setupRecyclerView() {
        mBinding?.rvTranscript?.apply {
            layoutManager = LinearLayoutManager(this@AgentChatActivity).apply {
                reverseLayout = false
            }
            adapter = transcriptAdapter
            itemAnimator = null
            addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    super.onScrollStateChanged(recyclerView, newState)
                    when (newState) {
                        RecyclerView.SCROLL_STATE_IDLE -> {
                            // Check if at bottom when scrolling stops
                            isScrollBottom = !recyclerView.canScrollVertically(1)
                            if (isScrollBottom) {
                                autoScrollToBottom = true
                                isScrollBottom = true
                            }
                        }

                        RecyclerView.SCROLL_STATE_DRAGGING -> {
                            // When user actively drags
                            autoScrollToBottom = false
                        }
                    }
                }

                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    super.onScrolled(recyclerView, dx, dy)
                    // Show button when scrolling up a significant distance
                    if (dy < -50) {
                        if (recyclerView.canScrollVertically(1)) {
                            autoScrollToBottom = false
                        }
                    }
                }
            })
        }
    }

    private fun observeUiState() {
        lifecycleScope.launch {
            viewModel.uiState.collect { state ->
                mBinding?.apply {
                    // Update button visibility based on connection state
                    val isConnected = state.connectionState == AgentChatViewModel.ConnectionState.Connected
                    val isConnecting = state.connectionState == AgentChatViewModel.ConnectionState.Connecting
                    val isIdle = state.connectionState == AgentChatViewModel.ConnectionState.Idle

                    // Show/hide views based on connection state
                    scrollView.visibility = if (isConnected) View.GONE else View.VISIBLE
                    cardTranscript.visibility = if (isConnected) View.VISIBLE else View.GONE
                    llControls.visibility = if (isConnected) View.VISIBLE else View.GONE
                    // Button state is managed by ChannelInputView

                    // Update mute button UI
                    btnMute.setImageResource(
                        if (state.isMuted) R.drawable.ic_mic_off else R.drawable.ic_mic
                    )

                    // Auto collapse avatar view when agent stops
                    if (isIdle && isAvatarExpanded) {
                        isAvatarExpanded = false
                        updateAvatarViewSize()
                    }
                }
            }
        }

        // Observe agent state
        lifecycleScope.launch {
            viewModel.agentState.collect { agentState ->
                mBinding?.apply {
                    agentState?.let {
                        // Update agent status text using state.value
                        tvAgentStatus.text = it.value
                    } ?: run {
                        // Agent state is null, show default text
                        tvAgentStatus.text = "Unknown"
                    }
                }
            }
        }
    }

    private fun observeTranscriptList() {
        lifecycleScope.launch {
            viewModel.transcriptList.collect { transcriptList ->
                // Update transcript list
                transcriptAdapter.submitList(transcriptList)
                if (autoScrollToBottom) {
                    scrollToBottom()
                }
            }
        }
    }

    private fun observeDebugLogs() {
        lifecycleScope.launch {
            viewModel.debugLogList.collect { logList ->
                mBinding?.apply {
                    // Update log text with all logs, separated by newlines
                    tvLog.text = logList.joinToString("\n").ifEmpty { "log" }
                    // Auto scroll to bottom
                    tvLog.post {
                        val scrollView = tvLog.parent as? android.widget.ScrollView
                        scrollView?.fullScroll(android.view.View.FOCUS_DOWN)
                    }
                }
            }
        }
    }

    /**
     * Observe avatar joined state to setup remote video
     */
    private fun observeAvatarJoined() {
        lifecycleScope.launch {
            viewModel.avatarJoined.collect { joined ->
                if (joined) {
                    setupAvatarVideo()
                } else {
                    hideAvatarVideo()
                }
            }
        }
    }

    /**
     * Setup avatar video view when avatar joins the channel
     */
    private fun setupAvatarVideo() {
        mBinding?.apply {
            // Setup remote video rendering using the SurfaceView in layout
            viewModel.setupRemoteVideo(avatarVideoView)

            // Show avatar video container
            avatarVideoContainer.visibility = View.VISIBLE
            isAvatarExpanded = false
            updateAvatarViewSize()
        }
    }

    /**
     * Hide avatar video when disconnected
     */
    private fun hideAvatarVideo() {
        mBinding?.apply {
            avatarVideoContainer.visibility = View.GONE
            isAvatarExpanded = false
            updateAvatarViewSize()
        }
    }

    /**
     * Toggle avatar video size between small and expanded
     */
    private fun toggleAvatarSize() {
        isAvatarExpanded = !isAvatarExpanded
        updateAvatarViewSize()
    }

    /**
     * Update avatar video view size based on expanded state
     * When expanded: cover the entire transcript area
     * When collapsed: small window at top-right corner of transcript area
     */
    private fun updateAvatarViewSize() {
        mBinding?.let { binding ->
            val container = binding.avatarVideoContainer
            val params = container.layoutParams as ConstraintLayout.LayoutParams
            val density = resources.displayMetrics.density

            if (isAvatarExpanded) {
                // Expanded: match cardTranscript size (cover entire transcript area)
                params.width = 0  // match_constraint
                params.height = 0  // match_constraint
                params.topToTop = binding.cardTranscript.id
                params.bottomToBottom = binding.cardTranscript.id
                params.startToStart = binding.cardTranscript.id
                params.endToEnd = binding.cardTranscript.id
                params.topMargin = 0
                params.marginEnd = 0
                params.marginStart = 0
                params.bottomMargin = 0
                // Corner radius is handled by drawable background
            } else {
                // Collapsed: small window at top-right corner
                params.width = (avatarViewSmallWidth * density).toInt()
                params.height = (avatarViewSmallHeight * density).toInt()
                params.topToTop = binding.cardTranscript.id
                params.bottomToBottom = ConstraintLayout.LayoutParams.UNSET
                params.startToStart = ConstraintLayout.LayoutParams.UNSET
                params.endToEnd = binding.cardTranscript.id
                val margin = (8 * density).toInt()
                params.topMargin = margin
                params.marginEnd = margin
                // Corner radius is handled by drawable background
            }
            container.layoutParams = params
        }
    }

    /**
     * Load saved channel name and UIDs and fill input fields
     */
    private fun loadSavedUIDs() {
        val (savedChannelName, uids) = viewModel.loadSavedChannelNameAndUIDs()
        val (savedUserId, savedAgentUid, savedAvatarUid) = uids
        mBinding?.channelInputView?.apply {
            loadSavedChannelName(savedChannelName)
            loadSavedUIDs(savedUserId, savedAgentUid, savedAvatarUid)
        }
    }
    
    /**
     * Scroll RecyclerView to the bottom to show latest transcript
     */
    private fun scrollToBottom() {
        mBinding?.rvTranscript?.apply {
            val lastPosition = transcriptAdapter.itemCount - 1
            if (lastPosition < 0) return

            stopScroll()
            val layoutManager = layoutManager as? LinearLayoutManager ?: return

            // Use single post call to handle all scrolling logic
            post {
                layoutManager.scrollToPosition(lastPosition)

                // Handle extra-long messages that exceed viewport height
                val lastView = layoutManager.findViewByPosition(lastPosition)
                if (lastView != null && lastView.height > height) {
                    val offset = height - lastView.height
                    layoutManager.scrollToPositionWithOffset(lastPosition, offset)
                }

                isScrollBottom = true
            }
        }
    }
}

/**
 * Adapter for displaying transcript list with different view types for USER and AGENT
 */
class TranscriptAdapter : ListAdapter<Transcript, RecyclerView.ViewHolder>(TranscriptDiffCallback()) {

    companion object {
        private const val VIEW_TYPE_USER = 0
        private const val VIEW_TYPE_AGENT = 1
    }

    override fun getItemViewType(position: Int): Int {
        return when (getItem(position).type) {
            TranscriptType.USER -> VIEW_TYPE_USER
            TranscriptType.AGENT -> VIEW_TYPE_AGENT
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            VIEW_TYPE_USER -> {
                UserViewHolder(ItemTranscriptUserBinding.inflate(LayoutInflater.from(parent.context), parent, false))
            }

            VIEW_TYPE_AGENT -> {
                AgentViewHolder(ItemTranscriptAgentBinding.inflate(LayoutInflater.from(parent.context), parent, false))
            }

            else -> throw IllegalArgumentException("Unknown view type: $viewType")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val transcript = getItem(position)
        when (holder) {
            is UserViewHolder -> holder.bind(transcript)
            is AgentViewHolder -> holder.bind(transcript)
        }
    }

    /**
     * ViewHolder for USER transcript items
     */
    class UserViewHolder(private val binding: ItemTranscriptUserBinding) : RecyclerView.ViewHolder(binding.root) {
        private val tvType: TextView = binding.tvTranscriptType
        private val tvText: TextView = binding.tvTranscriptText
        private val tvStatus: TextView = binding.tvTranscriptStatus

        fun bind(transcript: Transcript) {
            // Set transcript type badge with green color for USER
            tvType.text = "USER"
            ContextCompat.getDrawable(binding.root.context, R.drawable.bg_type_badge)?.let { drawable ->
                drawable.setTint("#10B981".toColorInt())
                tvType.background = drawable
            }

            // Set transcript text
            tvText.text = transcript.text.ifEmpty { "(empty)" }

            // Set transcript status with appropriate color
            val (statusText, statusColor) = when (transcript.status) {
                TranscriptStatus.IN_PROGRESS -> "IN PROGRESS" to "#FF9800".toColorInt()
                TranscriptStatus.END -> "END" to "#4CAF50".toColorInt()
                TranscriptStatus.INTERRUPTED -> "INTERRUPTED" to "#F44336".toColorInt()
                TranscriptStatus.UNKNOWN -> "UNKNOWN" to "#9E9E9E".toColorInt()
            }
            tvStatus.text = statusText
            tvStatus.setTextColor(statusColor)
        }
    }

    /**
     * ViewHolder for AGENT transcript items
     */
    class AgentViewHolder(private val binding: ItemTranscriptAgentBinding) : RecyclerView.ViewHolder(binding.root) {
        private val tvType: TextView = binding.tvTranscriptType
        private val tvText: TextView = binding.tvTranscriptText
        private val tvStatus: TextView = binding.tvTranscriptStatus

        fun bind(transcript: Transcript) {
            // Set transcript type badge with indigo color for AGENT
            tvType.text = "AGENT"
            ContextCompat.getDrawable(binding.root.context, R.drawable.bg_type_badge)?.let { drawable ->
                drawable.setTint("#6366F1".toColorInt())
                tvType.background = drawable
            }

            // Set transcript text
            tvText.text = transcript.text.ifEmpty { "(empty)" }

            // Set transcript status with appropriate color
            val (statusText, statusColor) = when (transcript.status) {
                TranscriptStatus.IN_PROGRESS -> "IN PROGRESS" to "#FF9800".toColorInt()
                TranscriptStatus.END -> "END" to "#4CAF50".toColorInt()
                TranscriptStatus.INTERRUPTED -> "INTERRUPTED" to "#F44336".toColorInt()
                TranscriptStatus.UNKNOWN -> "UNKNOWN" to "#9E9E9E".toColorInt()
            }
            tvStatus.text = statusText
            tvStatus.setTextColor(statusColor)
        }
    }

    private class TranscriptDiffCallback : DiffUtil.ItemCallback<Transcript>() {
        override fun areItemsTheSame(oldItem: Transcript, newItem: Transcript): Boolean {
            return oldItem.turnId == newItem.turnId && oldItem.type == newItem.type
        }

        override fun areContentsTheSame(oldItem: Transcript, newItem: Transcript): Boolean {
            return oldItem == newItem
        }
    }
}


